write/write.c

```diff
@@ -0,0 +1,212 @@
+		histtya = argv[2];
+	passw = getpwuid(geteuid());
+	if (passw) {
+		for (i = 0; c = passw->pw_name[i]; i++) me[i]=c;
+		me[(i + 1)] = '\0'; /* sender initials */
+	} else {
+		fprintf(stderr, "Can't open password database file\n");
+		exit(1);
+	}
+	mytty = ttyname(fileno(stderr));
+	if (mytty == NULL) {
+		fprintf(stderr, "Can't find your tty\n");
+		exit(1);
+	}
+	/* Get everything after /dev. */
+	for (i = 1; mytty[i] != '/'; i++);
+	mytty += (i + 1);
+	if (histtya[0] != '\0') {
+		strncpy(histty, "/dev/", 5);
+		strncat(histty, histtya, UT_LINESIZE);
+	}
+	setutxent();
+	while ((u = getutxent()) != NULL) {
+		if(him[0] != '-' || him[1] != 0)
+		switch (strncmp(him, u->ut_user, MAXNAMLEN)) {
+			case 0:
+				if (logcnt == 0 && histty[0]=='\0') {
+					strncpy(histty, "/dev/", 5);
+					strncat(histty, u->ut_line,
+							UT_LINESIZE);
+				} else {
+					otherplaces[places] = strdup(u->ut_line);
+					places++;
+				}
+				break;
+			default:
+				continue; /* nomat */
+		}
+		logcnt++;
+	}
+	endutxent();
+/* cont */
+	if (logcnt == 0 && histty[0] == '\0') {
+		fprintf(stderr, "%s not logged in.\n", him);
+		exit(1);
+	}
+	if (histtya[0] == '\0' && logcnt > 1) {
+		fprintf(stderr, "%s logged more than once\nwriting to %s\nother places where %s is at:\n",
+					him, histty, him);
+		for (i=0; i < places; i++)
+			fprintf(stderr, "%s\n", otherplaces[i]);
+	}
+	if (histty[0] == '\0') {
+		fprintf(stderr, him);
+		if (logcnt)
+			fprintf(stderr, " not on that tty\n");
+		else
+			fprintf(stderr, " not logged in\n");
+		exit(1);
+	}
+	signal(SIGALRM, timout);
+	alarm(5);
+	if ((tf = fopen(histty, "w")) == NULL)
+		goto perm;
+	alarm(0);
+	tfd = fileno(tf);
+	if (isatty(tfd) < 0) {
+		fprintf(stderr, "No such tty\n");
+		exit(1);
+	}
+	if (fstat(tfd, &stbuf) < 0)
+		goto perm;
+	/*
+	 * Check if the sender user's group (as per S_IWGRP)
+	 * or everyone (as per S_IWOTH) can write to
+	 * receiver's terminal.
+	 */
+	if ((stbuf.st_mode & (S_IWGRP|S_IWOTH)) == 0)
+		goto perm;
+	sigs(eof);
+	fprintf(tf, "Message from ");
+#ifdef interdata
+	fprintf(tf, "(Interdata) " );
+#endif
+	fprintf(tf, "%s %s...\n\7\7\7", me, mytty);
+	fflush(tf);
+	for(;;) {
+		char *buf = "";
+		char sbuf[128] = "";
+		i = read(0, sbuf, 128);
+		buf = ssafe(sbuf);
+
+		if (i <= 0)
+			eof();
+		switch (buf[0]) {
+			case '!':
+				buf[i] = 0;
+				ex(buf);
+				continue;
+			default:
+				break;
+		}
+		write(tfd, buf, i);
+	}
+
+perm:
+	fprintf(stderr, "Permission denied\n");
+	exit(1);
+}
+
+void timout(void) {
+	fprintf(stderr, "Timeout opening his tty\n");
+	exit(1);
+}
+
+void eof(void) {
+	fprintf(tf, "EOF\n");
+	exit(0);
+}
+
+int ex(char *bp) {
+	int i = 0;
+
+	sigs(SIG_IGN);
+	i = fork();
+	if (i < 0) {
+		fprintf(stderr, "Try again\n");
+		goto out;
+	}
+	if (i == 0) {
+		sigs((int (*)())0);
+		execl(SHELL, "sh", "-c", bp+1, 0);
+		exit(0);
+	}
+	while(wait((int *)NULL) != i)
+		;
+	printf("!\n");
+out:
+	sigs(eof);
+}
+
+int sigs(void (*sig)(int)) {
+	int i = 0;
```

Comment:
It would be better to declare `i` inside the loop

Reply:
As I said, unless we can fix the ``sigs()`` declaration for C99, this won't be possible because this is compiled as C89.

Reply:
funny

Reply:
Ported it to C99, but haven't declared i inside the loop because of the code style itself.